### ISSUE TITLE
Obfuscated gradients give a false sense of security: circumventing defenses to adversarial examples

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#371]
Obfuscated gradients give a false sense of security: circumventing defenses to adversarial examples<br>
<br>
https://ift.tt/2MRu8uw<br>
<br>
adriancolyer<br>
<br>
<div>
<div class="entry clear">
<p><a href="https://arxiv.org/abs/1802.00420">Obfuscated gradients give a false sense of security: circumventing defenses to adversarial examples</a> Athalye et al., <em>ICML&rsquo;18</em></p>
<p>There has been a lot of back and forth in the research community on adversarial attacks and defences in machine learning. Today&rsquo;s paper examines a number of recently proposed defences and shows that most of them rely on forms of <em>gradient masking</em>. The authors develop attack techniques to overcome such defences, and 9 analyse defences from ICLR 2018 claiming to protect against white-box attacks. 7 of these turn out to rely on obfuscated gradients, and 6 of these fall to the new attacks (and the other one partially succumbs). Athalye et al. won a best paper award at ICML&rsquo;18 for this work.</p>
<p>One of the great things about work on adversarial attacks and defences, as we&rsquo;ve looked at before, is that they illuminate the strengths and weaknesses of current technology. Depending on the threat model you choose, for my own part I&rsquo;m currently of the opinion that we&rsquo;re unlikely to find a robust adversarial defence without a more radical re-think of how we&rsquo;re doing image classification. If we&rsquo;re talking about the task of &lsquo;find an image that doesn&rsquo;t fool a human, but does fool a neural network&rsquo; then I think there will always be gaps to exploit until we incorporate more human-like (e.g., higher level semantic understanding) reasoning into our classifiers. Not that the human system itself can&rsquo;t be fooled of course (e.g., optical illusions).</p>
<h3>Breaking gradient descent</h3>
<blockquote>
<p>A defense is said to cause gradient masking if it &ldquo;does not have useful gradients&rdquo; for generating adversarial examples; gradient masking is known to be an incomplete defense to adversarial examples. Despite this, we observe that 7 of the ICLR 2018 defenses rely on this effect.</p>
</blockquote>
<p>Some defenses break gradient descent deliberately, others may do it unintentionally. Here are five clues that something isn&rsquo;t right:</p>
<ol>
<li>One-step attacks perform better than iterative attacks. Iterative attacks are strictly stronger, so this shouldn&rsquo;t be the case. If single-step methods are working better, it&rsquo;s a sign the iterative attack is becoming stuck at a local minimum.</li>
<li>Black-box attacks work better than white-box attacks. The black-box threat model is a strict subset of white-box attacks, so white-box attacks should perform better. When a defense obfuscates gradients, then black-box attacks (which don&rsquo;t use it) often perform better. (Since <a href="https://blog.acolyer.org/2017/02/28/when-dnns-go-wrong-adversarial-examples-and-what-we-can-learn-from-them/">practical black-box attacks are possible</a> and we can also find e.g. <a href="https://blog.acolyer.org/2017/09/12/universal-adversarial-perturbations/">universal adversarial perturbations</a> the utility of a defense that excludes such attack modes seems rather limited anyway to me).</li>
<li>Unbounded attacks do not reach 100% success. With unbounded distortion, any classifier should eventually fail. An attack that doesn&rsquo;t achieve this should be improved (i.e., it&rsquo;s a weak attack, not necessarily a strong defense).</li>
<li>Random sampling finds adversarial examples (where a gradient-based attack does not).</li>
<li>Increasing the distortion bound does not increase success. We expect a monotonically increasing attack success rate with increasing distortion bound.</li>
</ol>
<p>Three ways that a defense might break gradient descent are shattering, stochastic gradients, and exploding &amp; vanishing gradients:</p>
<ul>
<li>
<em>Shattered gradients</em> are caused when a defense is &ldquo;non-differentiable, introduces number instability, or otherwise causes a gradient to be nonexistent or incorrect.&rdquo;</li>
<li>
<em>Stochastic gradients</em> are a result of randomization &ndash; either introduced in the network itself, or in the inputs being fed to the network</li>
<li>
<em>Exploding and vanishing gradients</em> are often caused by defenses that consist of multiple iterations of neural network evaluation, feeding the output of one computation into the next.</li>
</ul>
<h3>Overcoming masked gradients</h3>
<p>Shattered gradients can be overcome using a technique the authors call &lsquo;Backward Pass Differentiable Approximation&rsquo; (BPDA). Think of a secured neural network (i.e., one that has been hardened against adversarial attacks) as being a composition of some hardening function <img src="https://s0.wp.com/latex.php?latex=g%28%5Cdot%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="g(\dot)" title="g(\dot)" class="latex"> and a regular pre-trained classifier <img src="https://s0.wp.com/latex.php?latex=f%28%5Cdot%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="f(\dot)" title="f(\dot)" class="latex"> such that the hardened classifier <img src="https://s0.wp.com/latex.php?latex=%5Chat%7Bf%7D%28x%29+%3D+f%28g%28x%29%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\hat{f}(x) = f(g(x))" title="\hat{f}(x) = f(g(x))" class="latex">. For example, <img src="https://s0.wp.com/latex.php?latex=g&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="g" title="g" class="latex"> might be a denoising function designed to remove perturbations introduced by an adversary. If <img src="https://s0.wp.com/latex.php?latex=g&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="g" title="g" class="latex"> itself is smooth and differentiable we can often compute gradients through the combined network <img src="https://s0.wp.com/latex.php?latex=%5Chat%7Bf%7D&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="\hat{f}" title="\hat{f}" class="latex">. Even if it <em>isn&rsquo;t</em> differentiable, we know that by construction in a local area it is close to the identify function (i.e. <img src="https://s0.wp.com/latex.php?latex=g%28x%29+%5Capprox+x&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="g(x) \approx x" title="g(x) \approx x" class="latex"> ). So we conduct an attack by performing forward propagation through the neural network as usual, but replacing <img src="https://s0.wp.com/latex.php?latex=g%28%5Cdot%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="g(\dot)" title="g(\dot)" class="latex"> with the identity function (or even simpler, just the evaluation of <img src="https://s0.wp.com/latex.php?latex=g%28x%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="g(x)" title="g(x)" class="latex"> ) during backpropagation. The full (general) version of BPDA operates by finding and replacing any non-differentiable layer <img src="https://s0.wp.com/latex.php?latex=f%5Ei%28%5Cdot%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="f^i(\dot)" title="f^i(\dot)" class="latex"> with a differentiable approximation <img src="https://s0.wp.com/latex.php?latex=g%28x%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="g(x)" title="g(x)" class="latex"> such that <img src="https://s0.wp.com/latex.php?latex=g%28x%29+%5Capprox+f%5Ei%28x%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="g(x) \approx f^i(x)" title="g(x) \approx f^i(x)" class="latex">. As before, we use the vanilla network on the forward pass, and substitute <img src="https://s0.wp.com/latex.php?latex=g&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="g" title="g" class="latex"> on the backward pass.</p>
<p>To attack a network that relies on stochastic gradients, it is necessary to estimate the gradient of the stochastic function. &lsquo;<a href="https://arxiv.org/abs/1707.07397">Expectation over Transformation</a>&rsquo; (EOT) can be used to compute the gradient over an expected transformation to an input. Given a transformation function <img src="https://s0.wp.com/latex.php?latex=t%28%5Cdot%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="t(\dot)" title="t(\dot)" class="latex"> EOT optimises the expectation over <img src="https://s0.wp.com/latex.php?latex=f%28t%28x%29%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="f(t(x))" title="f(t(x))" class="latex">. &ldquo;<em>The optimization problem can be solved through gradient descent&hellip; differentiating through the classifier and transformation, and approximating the expectation with samples at each gradient descent step</em>.&rdquo;</p>
<p>Vanishing or exploding gradients can be addressed through <em>reparameterization</em>. Suppose differentiating <img src="https://s0.wp.com/latex.php?latex=f%28g%28x%29%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="f(g(x))" title="f(g(x))" class="latex"> leads to a problem. We can introduce a new differentiable function <img src="https://s0.wp.com/latex.php?latex=h%28%5Cdot%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="h(\dot)" title="h(\dot)" class="latex"> such that <img src="https://s0.wp.com/latex.php?latex=x+%3D+h%28x%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="x = h(x)" title="x = h(x)" class="latex"> and <img src="https://s0.wp.com/latex.php?latex=g%28h%28z%29%29+%3D+h%28z%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="g(h(z)) = h(z)" title="g(h(z)) = h(z)" class="latex"> for all <img src="https://s0.wp.com/latex.php?latex=z&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="z" title="z" class="latex">. We can now compute gradients through <img src="https://s0.wp.com/latex.php?latex=f%28h%28z%29%29&amp;bg=ffffff&amp;fg=333333&amp;s=0" alt="f(h(z))" title="f(h(z))" class="latex"> and circumvent the defense. (&sect;4.3).</p>
<h3>How the ICLR 2018 defenses stack up</h3>
<p>Section 5 of the paper works through nine defenses from ICRL 2018 that claim robustness in a white-box threat model. Seven of the nine turn out to rely on some form of obfuscated gradients. For these the authors use the techniques described above to successfully attack them.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/obfuscated-gradients-table-1.jpeg?w=520"></p>
<p>Thermometer encoding is an example of a defense that relies on gradient shattering at its core. Image pixels are mapped to <em>l</em>-dimensional vectors such as &lsquo;111110000&rsquo;, where the number of leading 1&rsquo;s indicates the temperature (colour value) of a pixel. A BPDA attack reduces model accuracy to 30% (worse than if thermometer encoding were not used at all!).</p>
<p>Examples of transformation based defenses include image cropping and rescaling, bit-depth reduction, JPEG compression, randomly dropping pixels and replacing them using total variance minimization, and image quilting (reconstructing images by replaces small patches with patches taken from &lsquo;clean&rsquo; images).</p>
<blockquote>
<p>It is possible to bypass each defense independently (and ensembles of defenses usually are not much stronger than the strongest sub-component)&hellip; with our attack, we achieve 100% targeted attack success rate and accuracy drops to 0% for the strongest defense under the smallest perturbation budget&hellip;</p>
</blockquote>
<p>Image cropping and rescaling are circumvented using EOT. Bit depth reduction and JPEG compression are circumvented using BPDA and the identity function on the backward pass. Random pixel replacement and quilting are defeated using a combination of EOT and BPDA.</p>
<p>PixelDefend and Defense-GAN both use generators to project input images back onto a data manifold before feeding them into a classifier. Informally, this is intended to &lsquo;purify&rsquo; them by avoiding low-probability regions in the data distribution.</p>
<p>PixelDefend can be defeated using BPDA, and DefenseGAN can also be evaded with BPDA, though only at a 45% success rate.</p>
<p>See section 5 in the paper for more details and discussion of the other defenses.</p>
<blockquote>
<p>&hellip; we hope that future work will be able to avoid relying on obfuscated gradients (and other methods that only prevent gradient descent-based attacks) for perceived robustness, and use our evaluation approach to detect when this occurs.</p>
</blockquote>
</div>
</div>
<br>
<br>
<br>
<br>
BigData,AI|ML|DM|IR|PR<br>
<br>
via the morning paper https://ift.tt/1WJ1E4W<br>
<br>
August 15, 2018 at 01:00PM
